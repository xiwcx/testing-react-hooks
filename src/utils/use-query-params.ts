import { useCallback, useMemo } from "react";
import { useHistory } from "react-router-dom";
import { parse, stringify } from "query-string";

export function useQueryParams() {
  const history = useHistory();

  const parsedQueryParams = useMemo(
    () => parse(history.location.search),
    [history.location.search]
  );

  const clearParams = useCallback(
    () => history.push({ pathname: history.location.pathname }),
    [history]
  );

  const setParam = useCallback(
    (key: string, value: unknown) =>
      history.push({
        search: stringify({ ...parsedQueryParams, [key]: value }),
      }),
    [parsedQueryParams, history]
  );

  return {
    clearParams,
    params: parsedQueryParams,
    setParam,
  };
}
