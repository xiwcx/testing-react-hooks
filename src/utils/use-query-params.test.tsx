import { renderHook } from "@testing-library/react-hooks";
import { useQueryParams } from "./use-query-params";
import { createMemoryHistory, MemoryHistory } from "history";
import { Router } from "react-router-dom";

const renderUseQueryParamsHook = (history: MemoryHistory) =>
  renderHook(() => useQueryParams(), {
    wrapper: ({ children }) => <Router history={history}>{children}</Router>,
  });

describe("useQueryParams hook", () => {
  test("setParam works as expected", () => {
    const history = createMemoryHistory();
    const pushSpy = jest.spyOn(history, "push");
    const { result } = renderUseQueryParamsHook(history);

    result.current.setParam("foo", "bar");

    expect(pushSpy).toHaveBeenCalledTimes(1);
    expect(pushSpy).toHaveBeenCalledWith({ search: "foo=bar" });
    expect(history.location.search).toEqual("?foo=bar");
  });

  test.each([
    ["no params", "/"],
    ["no params", "/?foo=bar"],
  ])("params are retrieved correctly: %s", (_, initialPath) => {
    const history = createMemoryHistory();
    history.push(initialPath);
    const { result } = renderUseQueryParamsHook(history);

    expect(result.current.params).toMatchSnapshot();
  });

  test.each([
    ["root", "/", "/"],
    ["root", "/?foo=bar", "/"],
    ["root", "/test/?foo=bar", "/test/"],
  ])(
    "clearParams works as expected: %s",
    (_, initialPath, expectedPathname) => {
      const history = createMemoryHistory();
      history.push(initialPath);
      const pushSpy = jest.spyOn(history, "push");
      const { result } = renderUseQueryParamsHook(history);

      result.current.clearParams();

      expect(pushSpy).toHaveBeenCalledTimes(1);
      expect(pushSpy).toHaveBeenCalledWith({ pathname: expectedPathname });
      expect(history.location.pathname).toEqual(expectedPathname);
      expect(history.location.search).toEqual("");
    }
  );

  test.each([
    ["root", "/", "foo", "bar", "?foo=bar"],
    ["root", "/?foo=bar", "foo", "bar", "?foo=bar"],
    ["root", "/?foo=bar", "bar", "foo", "?bar=foo&foo=bar"],
  ])(
    "clearParams works as expected: %s",
    (_, initialPath, key, value, expectedPathname) => {
      const history = createMemoryHistory();
      history.push(initialPath);
      const { result } = renderUseQueryParamsHook(history);

      result.current.setParam(key, value);

      expect(history.location.search).toEqual(expectedPathname);
    }
  );
});
