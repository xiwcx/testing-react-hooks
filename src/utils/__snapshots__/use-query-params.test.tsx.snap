// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useQueryParams hook params are retrieved correctly: no params 1`] = `Object {}`;

exports[`useQueryParams hook params are retrieved correctly: no params 2`] = `
Object {
  "foo": "bar",
}
`;
